### PR TITLE
CrudeMortality

### PR DESCRIPTION
**Pull request by rimhajal**
the cmp.rel function on R outputs a causespecific mortality rate and a population one. this is (admittedly messy) close to the R code but the population rate is still not the right one

### CODE REVIEW

**Comment by lrnv** [comment]
A few leads:

1) I see the fit function actually does everything and that the `CrudeMortality()` function does nothing. This is not coherent with the other estimators right ? In the others, the fit functon is mostly dealing with the interface, and the `MyEstimator()` function is dealing with the rest, right ? More precisely, does this really need another `fit()` functin , cant we recycle the interface we already have? 
2) I see this crude mortality uses EdererII \Lambda function.  Would it make sense theoretically to allow other NPNSMethods there ? If so, then we could make it a type parameter as `CurdeMortality{M} where M<:NPNSEMethod`, and of course document the options correctly. 
3) Docs: we should include this crude mortality in the documentation, explaining what it is and how it works a bit like the rest, including in the examples. 
4) Tests: Test also the `CrudeMortality()` interface ? Test against R ?

**Comment by rimhajal** [comment]
1. The `CrudeMortality()` function is just for the struct and for the base.show function. And I didn't add it to the original fit function because it isn't a NPNSEMethod, so it wouldn't work with the type.
2. The EdererII method isn't ~really~ related, it isn't mentioned in the slides. But there are similarities in the calculations that's why I chose to use the function instead of redoing it.
3. On it.
4. The test compares with R and passes. I don't understand the first part about the interface?

**Comment by lrnv** [comment]
1. I thought that the `PoharPerme()`, `EdererI()`, `Ederer2()` and ``Hakulinen()` functions were already working as another-usable-but-not-recomended-interface ? If so, then I think `CrudeMortality()` should be the same. 
2. OK
3. Perfect
4. OK

**Comment by lrnv** [comment]
The conflicting file issue is because a previous PR deleted the theory.md file in the docs (i gathered it with another file). maybe you should: 

1. Undo last comit here
2.  copy/paste your modifications of the docs somewhere else (like a bloc note), 
3. Rebase this PR onto main 
4. re-do your docs modifications and comit them 
5. force-push your new branch to github

as we did together the other time

File: src/CrudeMortality.jl [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    grid            = mk_grid(resp[:,1],1)
+    num_pop         = zero(grid)
+    den_pop         = zero(grid)
+    num_excess      = zero(grid)
+    den_excess      = zero(grid)
+    num_variance    = zero(grid)
+
+    Λ!(EdererIIMethod, num_excess, den_excess, num_pop, den_pop, num_variance, resp[:,1], resp[:,2], df.age, df.year, rate_preds, rt, grid)
+    
+    ∂λₒ = num_excess ./ den_excess
```
Comment (lrnv) [inline src/CrudeMortality.jl:21]: should be \partial \Lambda_O and not \partial\lambda_O right ? (big lambda)

File: src/CrudeMortality.jl [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        ∂λₚ = 0.0
+
+        for j in 1:Tᵢ
+            ∂λₑ = (num_excess[j] ./ den_excess[j]) .- (num_pop[j] ./ den_pop[j])
+            Λₑ        += ∂λₑ * Sₒ[j]
+            ∂λₚ = num_pop[j] / den_pop[j]
+            causeSpec[j+1] = Λₑ
+            population[j+1] = population[j] + ∂λₚ * Sₒ[j]
+        end
+    end
```
Comment (lrnv) [inline src/CrudeMortality.jl:39]: This loop could be vectorized easily i think, will be much clearer what is happening no ?

**Comment by rimhajal** [comment]
I think we can absolutely do that, we wouldn't have anything to compare it to though. 
And in his pdf he doesn't really mention any method he just says that this is how he will estimate them.

File: src/CrudeMortality.jl (hunk 2 above)
Comment (rimhajal) [reply src/CrudeMortality.jl:51]: I am looking on how to do this but through a stand alone CrudeMortality function

**Comment by rimhajal** [comment]
oh wait ... I think I understood wrong. So we will add the other methods? We will just add in the docs that the Cronin-Feuer estimator uses EdererII?

**Comment by lrnv** [comment]
A good option would be to have the `NPNSEstimator` struct store not only `∂Λₑ` but also `∂Λₚ`, so that `∂Λₒ` can be obtained as the sum of the two, and then you can simply construct this crude mortality as 

```
X = fit(PoharPerme, .....)
Y = CrudeMortality(X)
```

since everything you need is already contained in X. This way you would not require the `fit` method to be redefined for the crude mortality, and the code will be much smaller, while retaining the possibility to get the crude mortality from any of the models. 

What do you think ? Do you see how this could be done or do you want me to commit a sketch ?

**Comment by lrnv** [comment]
> oh wait ... I think I understood wrong. So we will add the other methods? We will just add in the docs that the Cronin-Feuer estimator uses EdererII?

Yes I think so

**Comment by rimhajal** [comment]
> A good option would be to have the `NPNSEstimator` struct store not only `∂Λₑ` but also `∂Λₚ`, so that `∂Λₒ` can be obtained as the sum of the two, and then you can simply construct this crude mortality as
> 
> ```
> X = fit(PoharPerme, .....) # or Ederer or others
> Y = CrudeMortality(X)
> ```
> 
> since everything you need is already contained in X. This way you would not require the `fit` method to be redefined for the crude mortality, and the code will be much smaller, while retaining the possibility to get the crude mortality from any of the models.
> 
> What do you think ? Do you see how this could be done or do you want me to commit a sketch ?

Only issue with this is that we need the `∂Λₒ` before to get `∂Λₚ` and `∂Λₑ`. They are not the same values as the ones previously calculated.

**Comment by lrnv** [comment]
Well i see `∂Λₒ = num_excess ./ den_excess` in your code, so that is something that is already there i think

**Comment by lrnv** [comment]
And then you could have `fit(::Type{CreduMortality},args...) = CrudeMortality(fit(EdererII, args...))` to provide a default with the interface you already have.

**Comment by rimhajal** [comment]
> And then you could have `fit(::Type{CreduMortality},args...) = CrudeMortality(fit(EdererII, args...))` to provide a default with the interface you already have.

just saw this

**Comment by lrnv** [comment]
let me take a look at how it is right now

**Comment by rimhajal** [comment]
But how can it take the other methods?
Also am I adding too many things to the NPNSEstimator struct?

**Comment by lrnv** [comment]
This version looks good to me. Please proof-read the docs, and maybe add a graph to the example in the docs ? Was'nt this graph the whole point of this ?

**Comment by lrnv** [comment]
The fit function is not in the tests ? I thought that the tests passing was enough ?

**Comment by rimhajal** [comment]
the test passes but on the terminal it throws an error... 
also is the fit supposed to be only for EdererII?

**Comment by lrnv** [comment]
1) I have no error thrown on my terminal, can you at least copy/paste it here or just restart your julia session ? 
2) I guess that as the code currently is, yes, the `fit()` interface is only for the original ederer2 crude mortality, and the `CrudeMortality(fit())` interface is for the other ones. I tried a few others variants but this once seems the easiest to work with.

**Comment by lrnv** [comment]
Agreed, LGTM

**Pull request merged**